libqtile/widget/bluetooth_extra.py

```diff
@@ -0,0 +1,634 @@
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+from dbus_next.aio import MessageBus
+from dbus_next.constants import BusType
+from dbus_next.errors import DBusError, InterfaceNotFoundError
+
+from libqtile.command.base import expose_command
+from libqtile.log_utils import logger
+from libqtile.utils import create_task
+from libqtile.widget import base
+
+BLUEZ_SERVICE = "org.bluez"
+BLUEZ_DEVICE = "org.bluez.Device1"
+BLUEZ_ADAPTER = "org.bluez.Adapter1"
+BLUEZ_BATTERY = "org.bluez.Battery1"
+OBJECT_MANAGER_INTERFACE = "org.freedesktop.DBus.ObjectManager"
+PROPERTIES_INTERFACE = "org.freedesktop.DBus.Properties"
+
+
+def catch_error(msg):
+    """Decorator to catch DBusErrors and log a message."""
+
+    def _wrapper(func):
+        async def f(self):
+            try:
+                await func(self)
+            except DBusError:
+                # flake8 doesn't like formatting of log messages unless
+                # message will be inlcude in log.
+                # Mute the error here as it keeps code near.
+                logger.warning(msg.format(self=self))  # noqa: G001
+
+        return f
+
+    return _wrapper
+
+
+class _BluetoothBase:
+    """Base class with some common requirements for devices and adapters."""
+
+    def __init__(self, path, interface, properties_interface, widget):
+        self.path = path
+        self.interface = interface
+        self.widget = widget
+        self.properties = properties_interface
+        self.properties.on_properties_changed(self.properties_changed)
+        self._name = ""
+
+    def __repr__(self):
+        """Neater repr to help debugging."""
+        return f"<{self.__class__.__name__}: {self.name} ({self.path})>"
+
+    def __del__(self):
+        """Remove signal listener when garbage collected."""
+        try:
+            self.properties.off_properties_changed(self.properties_changed)
+        except RuntimeError:
+            pass
+
+    def properties_changed(self, _interface_name, _changed_properties, _invalidated_properties):
+        """Handler for properties_changed signal."""
+        create_task(self.update_props())
+
+    async def update_props(self):
+        raise NotImplementedError
+
```

Comment:
Is this method necessary? It seems it's only used where we know it won't error, so it might be fine just being defined on `BluetoothDevice`?

Reply:
It's just there in case someone wants to extend the widget in the future.